### PR TITLE
Remove validation that causes issue #1216

### LINKED ISSUE #1216: Cannot create a vdc with system default Sizing Policy (used as default compute policy) and VM Placement Policies
### Terraform Version
```
$ terraform version 
Terraform v1.7.2
on linux_amd64
+ provider registry.terraform.io/vmware/vcd v3.11.0
```

### Affected Resource(s)
- vcd_org_vdc

### Terraform Configuration Files
I can create a vdc not specifying vm policies:
```hcl
resource "vcd_org_vdc" "vdc" {
  . . .
  default_compute_policy_id = null
  vm_placement_policy_ids   = []
  vm_sizing_policy_ids      = []
  vm_vgpu_policy_ids        = []
  . . .
}
```
vcd creates a vdc with a system default sizing policy (empty) used as:
```
default_compute_policy_id
default_vm_sizing_policy_id (deprecated)
```
### Expected Behavior
Using the UI it is possible adding VM Placement Policies leaving the `default_compute_policy_id` and `default_vm_sizing_policy_id (deprecated)` untouched.
Using terraform, I cannot create a vdc with system default Sizing Policy (used as default compute policy) and VM Placement Policies, because I cannot specify the id of the  system default Sizing Policy (known after apply) and cannot be null.
I think it should be possible to set system default Sizing Policy as `default_compute_policy_id` even when a placement policy is added to the vdc.

As a workaround I created a dummy/empty VM Sizing Policy on cloud director, I assign it to the vdc and set it as default compute policy, so I can set placement policies.

I've added a comment on the closed issue https://github.com/vmware/terraform-provider-vcd/issues/1037 but I think it's better to open a new issue.

thanks

### PR DESCRIPTION
**Pull request by adambarreiro**
## Problem

As it is now, the VCD Provider requires that `default_compute_policy_id` is set when `vm_placement_policy_ids` or `vm_sizing_policy_ids` are specified.

This works for most of the cases, except one important one: When a VDC is created with several Placement policies assigned (`vm_placement_policy_ids`) but one wants to use the System default Sizing policy as default:

```hcl
resource "vcd_org_vdc" "TestAccVcdOrgVdcWithAllComputePolicies" {
  # ...omitted

  default_compute_policy_id = null # This will be computed to the System default policy, auto-generated from VCD
  vm_placement_policy_ids   = [
    vcd_vm_placement_policy.placement1.id,
    vcd_vm_placement_policy.placement2.id,
    vcd_vm_placement_policy.placement3.id,
    # ...
  ]
}
```

Applying a configuration like this will fail on runtime due to a validation the provider does in code:

```
!contains(vmComputePolicyIds, defaultPolicyId.(string))
```

## Fix

When there is no `default_compute_policy_id` set (or it is ""), ignore the validation:

```
defaultPolicyId.(string) != "" && !contains(vmComputePolicyIds, defaultPolicyId.(string))
```

## Tests

Added a new VDC definition that replicates this scenario in `TestAccVcdOrgVdcWithAllComputePolicies`

____

Closes #1216

### CODE REVIEW

**Comment by adambarreiro** [comment]
`vm` tests passed

File: .changes/v3.14.0/1313-bug-fixes.md [hunk 1]
```diff
@@ -0,0 +1,3 @@
+* Fix [Issue 1216](https://github.com/vmware/terraform-provider-vcd/issues/1216) in `vcd_org_vdc`
+  that would fail on creation when `vm_placement_policy_ids` are set but one does not want to declare a `default_compute_policy_id`
+  (to use the System default instead) [GH-1313]
```
Comment (lvirbalas) [inline .changes/v3.14.0/1313-bug-fixes.md:3]: Trying to simplify:

```suggestion
  which failed on creation when `vm_placement_policy_ids` were set but `default_compute_policy_id` was not declared (System default was used instead) [GH-1313]
```

File: vcd/resource_vcd_org_vdc.go [hunk 2]
```diff
@@ -972,8 +972,8 @@ func changeComputePoliciesAndDefaultId(d *schema.ResourceData, vcdClient *VCDCli
 	for _, attribute := range computePolicyAttributes {
 		vmComputePolicyIds = append(vmComputePolicyIds, convertSchemaSetToSliceOfStrings(d.Get(attribute).(*schema.Set))...)
 	}
-	if !contains(vmComputePolicyIds, defaultPolicyId.(string)) {
-		return fmt.Errorf("`default_compute_policy_id` %s is not present in any of `%v`", defaultPolicyId.(string), computePolicyAttributes)
+	if defaultPolicyId.(string) != "" && !contains(vmComputePolicyIds, defaultPolicyId.(string)) {
```
Comment (lvirbalas) [inline vcd/resource_vcd_org_vdc.go:975]: We could add a comment above this line, explaining the rationale and even mentioning the issue number.

**Review by lvirbalas — APPROVED** [review]
Few clarification suggestions inline.

**Review by Didainius — APPROVED** [review]

File: .changes/v3.14.0/1313-bug-fixes.md (hunk 1 above)
Comment (adambarreiro) [reply .changes/v3.14.0/1313-bug-fixes.md]: Changed, thanks!

File: vcd/resource_vcd_org_vdc.go (hunk 2 above)
Comment (adambarreiro) [reply vcd/resource_vcd_org_vdc.go:978]: Indeed, that would help. Added

**Comment by carmine73** [comment]
To avoid errors on reapply `vm_sizing_policy_ids` must be set to `null` or omitted (sorry for this comment, probably it is useful only for me).
```
resource "vcd_org_vdc" "TestAccVcdOrgVdcWithAllComputePolicies" {
  # ...omitted

  default_compute_policy_id = null # This will be computed to the System default policy, auto-generated from VCD
  vm_sizing_policy_ids      = null # cannot be [], otherwise leads to an error on re-apply
  vm_placement_policy_ids   = []
}
```

**Comment by carmine73** [comment]
@adambarreiro I have this objective: setting some sizing policies but I would like to have the system default policy as the default one.

my code:
```
## org-vdc
resource "vcd_org_vdc" "vdc" {
  count = local.vdc.required == true ? 1 : 0 # only if vdc is required

  . . .
  default_compute_policy_id = local.default_compute_policy_id
  vm_sizing_policy_ids      = length(local.vm_sizing_policy_ids) > 0 ? local.vm_sizing_policy_ids : null
  vm_placement_policy_ids   = try(local.vdc.placement_policy_ids, [])
  . . .
}
```
in my case `local.default_compute_policy_id` is 'null' and `local.vm_sizing_policy_ids` has 4 policies.

plan output:
```
  # vcd_org_vdc.vdc[0] will be created
  + resource "vcd_org_vdc" "vdc" {
      . . .
      + default_compute_policy_id        = (known after apply)
      + default_vm_sizing_policy_id      = (known after apply)
      . . .
      + vm_placement_policy_ids          = (known after apply)
      + vm_sizing_policy_ids             = [
          + "urn:vcloud:vdcComputePolicy:1c1bf120-....-f59e29e8e33c",
          + "urn:vcloud:vdcComputePolicy:4a102293-....-0da7973dbef5",
          + "urn:vcloud:vdcComputePolicy:82c1731b-....-f1237af0dce8",
          + "urn:vcloud:vdcComputePolicy:e3b95405-....-fd0448ad397b",
        ]
      + vm_vgpu_policy_ids               = (known after apply)
```
after apply:
```
$ terraform state show vcd_org_vdc.vdc[0]
# vcd_org_vdc.vdc[0]:
resource "vcd_org_vdc" "vdc" {
    . . .
    default_compute_policy_id        = "urn:vcloud:vdcComputePolicy:cf3a9fad-....-112ec862dc4a"
    default_vm_sizing_policy_id      = "urn:vcloud:vdcComputePolicy:cf3a9fad-....-112ec862dc4a"
    . . .
    vm_placement_policy_ids          = []
    vm_sizing_policy_ids             = [
        "urn:vcloud:vdcComputePolicy:1c1bf120-....-f59e29e8e33c",
        "urn:vcloud:vdcComputePolicy:4a102293-....-0da7973dbef5",
        "urn:vcloud:vdcComputePolicy:82c1731b-....-f1237af0dce8",
        "urn:vcloud:vdcComputePolicy:e3b95405-....-fd0448ad397b",
    ]
    vm_vgpu_policy_ids               = []
```
as you can se the `default_compute_policy_id` (id ending with `dc4a`) is not included in the `vm_sizing_policy_ids`, while on UI I can see the system default policy (there are 5 policies, not just 4)

So this is what I want, but if I apply again I got an error, because of this:

```
  # vcd_org_vdc.vdc[0] will be updated in-place
  ~ resource "vcd_org_vdc" "vdc" {
        . . .
      ~ vm_sizing_policy_ids             = [
          - "urn:vcloud:vdcComputePolicy:cf3a9fad-....-112ec862dc4a",
            # (4 unchanged elements hidden)
        ]
```
this is the error, that I understand.
I cannot set the id of the default policy without a data source, that cannot be done without knowing that is not the first apply.
```
Error: error assigning VM sizing policies to VDC: `default_compute_policy_id` urn:vcloud:vdcComputePolicy:cf3a9fad-....-112ec862dc4a is not present in any of `[vm_sizing_policy_ids vm_placement_policy_ids vm_vgpu_policy_ids]`
```
Is there a way I can solve this on a second apply?

**Comment by carmine73** [comment]
Hi @adambarreiro 
I'm trying to explain my case better.
My actual configuration code is:
```
resource "vcd_org_vdc" "vdc" {
    default_compute_policy_id        = null
    vm_placement_policy_ids          = []
    vm_sizing_policy_ids             = [
        "urn:vcloud:vdcComputePolicy:1c1bf120-....-f59e29e8e33c",
        "urn:vcloud:vdcComputePolicy:4a102293-....-0da7973dbef5",
        "urn:vcloud:vdcComputePolicy:82c1731b-....-f1237af0dce8",
        "urn:vcloud:vdcComputePolicy:e3b95405-....-fd0448ad397b",
    ]
    vm_vgpu_policy_ids               = []
}
```
this is my vdc after first apply:
```
# vcd_org_vdc.vdc[0]:
resource "vcd_org_vdc" "vdc" {
    default_compute_policy_id        = "urn:vcloud:vdcComputePolicy:33fbf269-....-0c883633953b" # system default compute policy id
    default_vm_sizing_policy_id      = "urn:vcloud:vdcComputePolicy:33fbf269-....-0c883633953b"
    id                               = "urn:vcloud:vdc:ba0ddca4-....-d277786a0e28"
    vm_placement_policy_ids          = []
    vm_sizing_policy_ids             = [
        "urn:vcloud:vdcComputePolicy:1c1bf120-....-f59e29e8e33c",
        "urn:vcloud:vdcComputePolicy:4a102293-....-0da7973dbef5",
        "urn:vcloud:vdcComputePolicy:82c1731b-....-f1237af0dce8",
        "urn:vcloud:vdcComputePolicy:e3b95405-....-fd0448ad397b",
    ]
    vm_vgpu_policy_ids               = []
}
```
I had a look inside terraform logs:
after vDC is created (I guess system default policy is created automatically by vcd) there is this PUT to update compute policies, which include the system default policy (kind of weird since I cannot put it in my code).
```
PUT https://my-vcd/api/admin/vdc/ba0ddca4-...-d277786a0e28/computePolicies
...
  <VdcComputePolicyReferences xmlns="http://www.vmware.com/vcloud/v1.5">
      <VdcComputePolicyReference href="https://my-vcd/cloudapi/2.0.0/vdcComputePolicies/urn:vcloud:vdcComputePolicy:e3b95405-....-fd0448ad397b"></VdcComputePolicyReference>
      <VdcComputePolicyReference href="https://my-vcd/cloudapi/2.0.0/vdcComputePolicies/urn:vcloud:vdcComputePolicy:4a102293-....-0da7973dbef5"></VdcComputePolicyReference>
      <VdcComputePolicyReference href="https://my-vcd/cloudapi/2.0.0/vdcComputePolicies/urn:vcloud:vdcComputePolicy:1c1bf120-....-f59e29e8e33c"></VdcComputePolicyReference>
      <VdcComputePolicyReference href="https://my-vcd/cloudapi/2.0.0/vdcComputePolicies/urn:vcloud:vdcComputePolicy:82c1731b-....-f1237af0dce8"></VdcComputePolicyReference>
      <VdcComputePolicyReference href="https://my-vcd/cloudapi/2.0.0/vdcComputePolicies/urn:vcloud:vdcComputePolicy:33fbf269-....-0c883633953b"></VdcComputePolicyReference>
  </VdcComputePolicyReferences>
```
So I do again plan (with no changes), I see this GET
```
GET https://my-vcd/cloudapi/2.0.0/vdcs/urn:vcloud:vdc:ba0ddca4-....-d277786a0e28/computePolicies?pageSize=128
```
that reads all 5 compute policies.
During (re)apply I got the error:
```
Error: error assigning VM sizing policies to VDC: `default_compute_policy_id` urn:vcloud:vdcComputePolicy:33fbf269-....-0c883633953b is not present in any of `[vm_sizing_policy_ids vm_placement_policy_ids vm_vgpu_policy_ids]`
```
I don't see errors on logs for API calls, I guess it's an error raised by vcd terraform provider, why is it not raised during first apply?
Is there a way to have the same behavior of first apply?

**to summarize, during vdc creation I would like to have**
* system default policy created
* more sizing policies added by me
* do not have errors on re-apply (this is the only missing part  today)

thanks.

**Comment by carmine73** [comment]
Hi @adambarreiro,
finally I've found a way to managed reapply on my use case
